### ISSUE TITLE
lookup function on cast map returns string cast null value

### ISSUE THREAD

**Issue by asvinours** [hashicorp/terraform#29171]
### Terraform Version

```bash
$ terraform --version
Terraform v1.0.2
```

### Terraform Configuration Files

```terraform
locals {
  my_var_1 = {
    my_key_1 = "bob"
    my_key_2 = "john"
  }

  my_var_2 = tomap({
    my_key_1 = "alice"
    my_key_2 = "julie"
  })
}
```

### Expected Behavior

I'm expecting to get the specified return value with no modification

### Actual Behavior

terraform is casting the specified `null` value as a string if the map object is cast as a map.
we found this because of assigning map objects to a variable using an if condition.
the condition is enforcing `tomap` casting on the map objects, and when using lookup the null
value is being cast to a string which is an invalid value.

for more context, this is the exact lookup causing issue for us: https://github.com/terraform-aws-modules/terraform-aws-alb/blob/v6.3.0/main.tf#L160

```
│ Error: Incorrect attribute value type
│ 
│   on .terraform/modules/alb/main.tf line 160, in resource "aws_lb_listener_rule" "https_listener_rule":
│  160:         authentication_request_extra_params = lookup(action.value, "authentication_request_extra_params", null)
│ 
│ Inappropriate value for attribute "authentication_request_extra_params":
│ map of string required.
```

### Steps to Reproduce
```
$ terraform console
> lookup(local.my_var_1, "lolipop", null)
null
> lookup(local.my_var_2, "lolipop", null)
tostring(null)
```

**Comment by jbardin** [comment]
Hi @asvinours,

Can you explain a little more about the actual problem you had? The behavior here looks correct, if `my_var_2` is a `map(string)`, then the only possible type it can return is a string, which is shown in your example. It does not mean the result cannot still be `null`, however that `null` does have a type of `string` in this the case of the map. 

I'm guessing from the error output you are ending up with an object containing mixed  types of some sort, but it would help if we had an isolated case to demonstrate the issue. 

Thanks!

**Comment by asvinours** [comment]
hey,

Let me show you the real world example, maybe I'm doing something wrong.

```terraform
locals {
  enable_cognito = true

  cognito_userpool_action = {
    type = "authenticate-cognito"

    user_pool_arn       = var.cognito_user_pool_arn
    user_pool_client_id = var.cognito_user_pool_client_id
    user_pool_domain    = var.cognito_user_pool_domain
  }

  base_listener_rule = {
    https_listener_index = 0
    priority             = 10

    conditions = [{
      host_headers = ["*.example.com"]
    }]
  }

  forward_action = {
    type               = "forward"
    target_group_index = 0
  }

  site_listener_rule = merge(
    local.base_listener_rule, 
    { 
      actions = var.enable_cognito ? [local.cognito_userpool_action, local.forward_action] : [local.forward_action]
    }
  )
}

module "alb" {
  source  = "terraform-aws-modules/alb/aws"
  version = "~> 6.0"

  name               = "my-alb"
  load_balancer_type = "application"

  vpc_id  = local.main_vpc
  subnets = local.alb_subnets

  target_groups = local.alb_target_groups

  security_groups = [local.alb_security_group]

  http_tcp_listeners = local.alb_http_tcp_listeners

  https_listeners = [
    local.site_listener_rule,
  ]
}
```

Using the terraform console, I can print the content of `local.site_listener_rule`:

```
{
  "actions" = tolist([
    tomap({
      "type" = "authenticate-cognito"
      "user_pool_arn" = "arn:aws:cognito-idp:region:1234:userpool/abcdefgh"
      "user_pool_client_id" = "abcdefgh"
      "user_pool_domain" = "auth.example.com"
    }),
    tomap({
      "target_group_index" = "0"
      "type" = "forward"
    }),
  ])
  "conditions" = [
    {
      "host_headers" = [
        "*.example.com",
      ]
    },
  ]
  "https_listener_index" = 0
  "priority" = 10
}
```

the issue I'm facing is, when using lookup on the map inside the action list, instead of returning a `null` value, it's returning a string that contains `null`

```
> lookup(local.site_listener_rule["actions"][0], "authentication_request_extra_params", null)
tostring(null)
```

if I remove the condition and set `site_listener_rule` using this code:

```terraform
locals {
  site_listener_rule = merge(
    local.base_listener_rule, 
    { 
      actions = [local.cognito_userpool_action, local.forward_action]
    }
  )
}
```

then the output of `local.site_listener_rule` is different (no `tomap` anymore):

```
{
  "actions" = [
    {
      "type" = "authenticate-cognito"
      "user_pool_arn" = "arn:aws:cognito-idp:us-east-1:805395716309:userpool/us-east-1_70702FWfd"
      "user_pool_client_id" = "65i2r86g2qrj2hl0o1rq6v8t55"
      "user_pool_domain" = "auth.renorun.com"
    },
    {
      "target_group_index" = 0
      "type" = "forward"
    },
  ]
  "conditions" = [
    {
      "host_headers" = [
        "*.renorun.ca",
      ]
    },
  ]
  "https_listener_index" = 0
  "priority" = 70
}
```

and the `lookup` now returns a real `null` value:

```
> lookup(local.site_listener_rule["actions"][0], "authentication_request_extra_params", null)
null
```

`authentication_request_extra_params` attribute of the `aws_lb_listener_rule` resource seems to accept `null` but reject `tostring(null)` value

**Comment by asvinours** [comment]
hey,

Let me show you the real world example, maybe I'm doing something wrong.

```terraform
locals {
  enable_cognito = true

  cognito_userpool_action = {
    type = "authenticate-cognito"

    user_pool_arn       = var.cognito_user_pool_arn
    user_pool_client_id = var.cognito_user_pool_client_id
    user_pool_domain    = var.cognito_user_pool_domain
  }

  base_listener_rule = {
    https_listener_index = 0
    priority             = 10

    conditions = [{
      host_headers = ["*.example.com"]
    }]
  }

  forward_action = {
    type               = "forward"
    target_group_index = 0
  }

  site_listener_rule = merge(
    local.base_listener_rule, 
    { 
      actions = var.enable_cognito ? [local.cognito_userpool_action, local.forward_action] : [local.forward_action]
    }
  )
}

module "alb" {
  source  = "terraform-aws-modules/alb/aws"
  version = "~> 6.0"

  name               = "my-alb"
  load_balancer_type = "application"

  vpc_id  = local.main_vpc
  subnets = local.alb_subnets

  target_groups = local.alb_target_groups

  security_groups = [local.alb_security_group]

  http_tcp_listeners = local.alb_http_tcp_listeners

  https_listeners = [
    local.site_listener_rule,
  ]
}
```

Using the terraform console, I can print the content of `local.site_listener_rule`:

```
{
  "actions" = tolist([
    tomap({
      "type" = "authenticate-cognito"
      "user_pool_arn" = "arn:aws:cognito-idp:region:1234:userpool/abcdefgh"
      "user_pool_client_id" = "abcdefgh"
      "user_pool_domain" = "auth.example.com"
    }),
    tomap({
      "target_group_index" = "0"
      "type" = "forward"
    }),
  ])
  "conditions" = [
    {
      "host_headers" = [
        "*.example.com",
      ]
    },
  ]
  "https_listener_index" = 0
  "priority" = 10
}
```

the issue I'm facing is, when using lookup on the map inside the action list, instead of returning a `null` value, it's returning a string that contains `null`

```
> lookup(local.site_listener_rule["actions"][0], "authentication_request_extra_params", null)
tostring(null)
```

if I remove the condition and set `site_listener_rule` using this code:

```terraform
locals {
  site_listener_rule = merge(
    local.base_listener_rule, 
    { 
      actions = [local.cognito_userpool_action, local.forward_action]
    }
  )
}
```

then the output of `local.site_listener_rule` is different (no `tomap` anymore):

```
{
  "actions" = [
    {
      "type" = "authenticate-cognito"
      "user_pool_arn" = "arn:aws:cognito-idp:region:1234:userpool/abcdefgh"
      "user_pool_client_id" = "abcdefgh"
      "user_pool_domain" = "auth.example.com"
    },
    {
      "target_group_index" = 0
      "type" = "forward"
    },
  ]
  "conditions" = [
    {
      "host_headers" = [
        "*.example.com",
      ]
    },
  ]
  "https_listener_index" = 0
  "priority" = 70
}
```

and the `lookup` now returns a real `null` value:

```
> lookup(local.site_listener_rule["actions"][0], "authentication_request_extra_params", null)
null
```

`authentication_request_extra_params` attribute of the `aws_lb_listener_rule` resource seems to accept `null` but reject `tostring(null)` value

**Comment by jbardin** [comment]
The `aws_lb_listener_rule` declares that `authentication_request_extra_params` is a `map(map(string))`, so a string value is not valid there, which is why the null string is rejected with the error `map of string required`. I think the module here could do better by specifying the structure of the `https_listener_rules`  rather than leaving it as an `any` type.

Since the problem here doesn't seem to be a bug in Terraform, I'm going to close the issue. Feel free to response here if you have any follow up terraform questions, but you may want to start a thread in [the community forum](https://discuss.hashicorp.com/) where there are more people familiar with the AWS modules and provider ready to help.

Thanks!

**Comment by asvinours** [comment]
I don't understand your answer.

> The aws_lb_listener_rule declares that authentication_request_extra_params is a map(map(string)), so a string value is not valid there, which is why the null string is rejected with the error map of string required

This is indeed the issue. But the module is asking `lookup` to return `null` not `null string`.

Are `null` and `tostring(null)` the same thing? it feels like one is null (as in "a value that represents absence or omission") and the other one is a string that contains the word "null".

**Comment by asvinours** [comment]
I did more tests

```terraform
locals {
  my_map = {
    my_key = {
      my_sub_key = "bob"
    }
  }

  my_other_map = tomap({
    my_key = tomap({
      my_sub_key = "bob"
    })
  })
}

module "my_module" {
  source = "./modules/test-null"

  my_variable = lookup(local.my_map, "lolipop", null)
}

module "my_2_module" {
  source = "./modules/test-null"

  my_variable = lookup(local.my_other_map, "lolipop", null)
}

module "my_3_module" {
  source = "./modules/test-null"

  my_variable = null
}

module "my_4_module" {
  source = "./modules/test-null"

  my_variable = tostring(null)
}
```

and here is the content of the `test-null` module:

```terraform
variable "my_variable" {
  type = map(map(string))
}

output "my_variable" {
  value = var.my_variable == null ? "that value is null" : "that value is not null"
}


output "my_variable_value" {
  value = var.my_variable
}
```

I get the following error:

```
╷
│ Error: Invalid value for module argument
│ 
│   on test.tf line 24, in module "my_2_module":
│   24:   my_variable = lookup(local.my_other_map, "lolipop", null)
│ 
│ The given value is not suitable for child module variable "my_variable" defined at modules/test-null/main.tf:1,1-23: incorrect map element type: map of string required.
╵
```

I can send `tostring(null)` value and it accepts it, but the return value of the lookup is not accepted.
I don't really understand why this is happening.

**Comment by jbardin** [comment]
The `tostring(null)` is the format used to represent a `null` value with a string type constraint, it does not represent a string value or a string with the word `"null"`. 

The problem in the latest example is that you have defined the input to the module as `map(string(string))`, but the value which is failing is of the type `map(string)`. Passing the entire map which is the same type will work

```
module "my_2_module" {
  source = "./modules/test-null"
  my_variable = local.my_other_map
}
```

The other example do work, because you are using objects where the lookup cannot determine an exact type, so the resulting value is an un-typed `null` which can pass for anything.
